### ISSUE TITLE
Loss explosion with DataParallel on WGAN models 

### ISSUE THREAD

**Issue by pietern** [pytorch/pytorch#19024]
See https://discuss.pytorch.org/t/huge-loss-with-dataparallel/40749 for the original report.

This seems to have regressed between 0.4.1 and 1.0 and needs to be debugged.

cc @mrshenli

**Comment by MKFMIKU** [comment]
In the Pytorch version 1.0.1-py3.7_cuda8.0.61_cudnn7.1.2_2, I face the same problem.

**Comment by MKFMIKU** [comment]
After testing, the problem still existed in the newest PyTorch version 1.1.0.dev20190417

**Comment by ailias** [comment]
Yes, I have also met this issue when I use pytorch==1.0.1 version,  but if I  use 0.4.1 version everything is fine. So what's wrong?  This problem waste some of my time for debugging my program.

**Comment by ailias** [comment]
![image](https://user-images.githubusercontent.com/10510469/56457779-22c25380-63b2-11e9-9ff3-8264ca9cc309.png)

Same code for different pytorch version while cause different loss.

**Comment by neale** [comment]
I can also confirm that this issue is in the latest build of 1.1.0. 
Cuda 10. Python 3.7.2.

**Comment by egurapha** [comment]
I also see this occurring for pytorch 1.0.1  on python 3.7.3, with cuda10.0.130, cudnn7.4.

**Comment by hukkelas** [comment]
I'm struggling with the same issue with pytorch 1.0

I've made a minimal example to reproduce this:
```python
import torch
from torch import nn
from torch.optim import Adam
import numpy as np
import os

torch.manual_seed(0)
np.random.seed(0)
torch.backends.cudnn.benchmark = True
torch.backends.cudnn.deterministic = True
torch.set_printoptions(precision=10)


discriminator1 = nn.Linear(1, 1, bias=True).cuda()

discriminator2 = nn.Linear(1, 1, bias=True)
discriminator2.load_state_dict(discriminator1.state_dict())
discriminator2 = nn.DataParallel(discriminator2).cuda()

optim1 = Adam(discriminator1.parameters())
optim2 = Adam(discriminator2.parameters())

def gradient_penalty(real_data, discriminator):
    real_data = real_data.clone()
    real_data.requires_grad=True
    logits = discriminator(real_data).sum()
    grad = torch.autograd.grad(
        outputs=logits,
        inputs=real_data,
        grad_outputs=torch.ones(logits.shape).cuda(),
        create_graph=True
    )[0]
    return grad.sum()

# Define input
x_fake = torch.randn((1, 1)).cuda()

# Discriminator 1
penalty = gradient_penalty(x_fake, discriminator1).sum()
optim1.zero_grad()
penalty.backward()
optim1.step()
y_out1 = discriminator1(x_fake)

# Discriminator 2
penalty = gradient_penalty(x_fake, discriminator2).sum()
optim2.zero_grad()
penalty.backward()
optim2.step()
y_out2 = discriminator2(x_fake)

print(y_out1 - y_out2)
```
Optimally, the difference between `y_out1` and `y_out2` should be 0.

Tested with the following NGC pytorch containers:

- [nvcr.io/nvidia/pytorch:18.09-py3](https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel_18.09.html). This is based on 0.4.1 official release. The difference is 0.0, no issue with gradient penalty.
- [nvcr.io/nvidia/pytorch:18.10-py3](https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel_18.11.html). This is based on a 1.0 preview (Master branch up to PR [11834](https://github.com/pytorch/pytorch/pull/11834) The difference is 0.0021787882, which is a significant difference.

I also checked the newest pytorch version. Still broken.

Hope this helps!

**Comment by thuyen** [comment]
It's a bug. `DataParallel` now does not handle double backward properly. In the above example `discriminator2` does not get any update. 

```
import torch
from torch import nn
from torch.optim import Adam
import numpy as np
import os

torch.backends.cudnn.benchmark = True
torch.backends.cudnn.deterministic = True
torch.set_printoptions(precision=10)


discriminator1 = nn.Linear(1, 1, bias=True).cuda()
discriminator2 = nn.Linear(1, 1, bias=True)
discriminator2.load_state_dict(discriminator1.state_dict())
discriminator2 = nn.DataParallel(discriminator2).cuda()

learning_rate = 0.5 # put any value here
optim1 = Adam(discriminator1.parameters(), lr=learning_rate)
optim2 = Adam(discriminator2.parameters(), lr=1000)

def gradient_penalty(real_data, discriminator):
    real_data = real_data.clone()
    real_data.requires_grad=True
    logits = discriminator(real_data).sum()
    grad = torch.autograd.grad(
        outputs=logits,
        inputs=real_data,
        grad_outputs=torch.ones(logits.shape).cuda(),
        create_graph=True
    )[0]
    return grad.sum()

# Define input
x_fake = torch.ones((1, 1)).cuda()

# Discriminator 1
penalty = gradient_penalty(x_fake, discriminator1).sum()
optim1.zero_grad()
penalty.backward()
optim1.step()

# Discriminator 2
penalty = gradient_penalty(x_fake, discriminator2).sum()
optim2.zero_grad()
penalty.backward()
optim2.step()

# this always prints something around 1e-7 regardlesss of learning_rate and initialization
# discriminator2 basically does not get any grad update
print(discriminator2.module.weight - discriminator1.weight - learning_rate)

```

**Comment by mdlockyer** [comment]
Okay. I think I'm running into this same issue from DistributedDataParallel as well. The loss chart above looks strikingly similar to what I'm seeing when using a gradient penalty (not the WGAN-GP style though). The loss has some noise to it, but it does't trend up or down. And the model just simply doesn't learn. I've been banging my head against my desk for over a month now trying to debug my architecture then I found this issue. 

After just reimplementing my model in TensorFlow (its a 1:1 implementation, including the ability to swap weights between TF and PyTorch and get the exact same outputs) and Viola! It trains perfectly, as it should have been. The training methodology is exactly the same between implementations as well. It seems second order derivatives are broken. At least when using multi-GPU training (I can't run my model on a single GPU to test it so I'm not sure). 

Any progress being made on this? It seems very important.

**Comment by hukkelas** [comment]
@mdlockyer, I'm not sure if it is the same issue with DistributedDataParallel. According to the [docs](https://pytorch.org/docs/stable/nn.html#torch.nn.parallel.DistributedDataParallel), DistributedDataParallel does not support torch.grad.autograd, which is required for gradient penalty

> Okay. I think I'm running into this same issue from DistributedDataParallel as well.

**Comment by gchanan** [comment]
@pietern @mrshenli any chance a fix gets into 1.2?

**Comment by gchanan** [comment]
assigning @pietern to figure out if we should target this for 1.2 or not.

**Comment by pietern** [comment]
@thuyen Many thanks for the repro script.

I'm running a bisect to figure out which commit introduced this.

**Comment by pietern** [comment]
I've bisected to #11611 being the culprit here.

@resistor @apaszke @ezyang Any ideas?

**Comment by pietern** [comment]
Assigning to @resistor for further investigation.

**Comment by thuyen** [comment]
@pietern A fix was attempted in #16888.

**Comment by pietern** [comment]
And the original issue for that fix at #16532. Let's continue the discussion there to consolidate.
